proto/all_test.go

```diff
@@ -1860,6 +1860,21 @@ func TestRequiredNotSetError(t *testing.T) {
 	}
 }
 
+func TestRequiredNotSetErrorWithBadWireTypes(t *testing.T) {
+	// Required field is properly a varint.
+	if err := Unmarshal([]byte{0x08, 0x00}, new(GoEnum)); err != nil {
+		t.Errorf("Unmarshal = %v, want nil", err)
+	}
+	// Required field is a fixed32 instead of varint.
+	if err := Unmarshal([]byte{0x0d, 0x00, 0x00, 0x00, 0x00}, new(GoEnum)); err == nil {
+		t.Errorf("Unmarshal = nil, want RequiredNotSetError")
+	}
+	// Required field has both varint and fixed32 (ignored).
```

Comment:
Same here.

// Bytes contain both varint and fixed32 for same required varint field.

Btw, given that both encoded fields are for key 1, shouldn't that supposed to be an error?

Reply:
> Btw, given that both encoded fields are for key 1, shouldn't that supposed to be an error?

Not according to the C++ implementation :)